pkgs/dart_tooling_mcp_server/lib/src/mixins/dtd.dart

```diff
@@ -0,0 +1,292 @@
+  void _resetDtd() {
+    _dtd = null;
+    _getDebugSessionsReady = false;
+  }
+
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) async {
+    registerTool(_connectTool, _connect);
+    registerTool(_screenshotTool, takeScreenshot);
+    return super.initialize(request);
+  }
+
+  /// Connects to the Dart Tooling Daemon.
+  FutureOr<CallToolResult> _connect(CallToolRequest request) async {
+    if (_dtd != null) {
+      return _dtdAlreadyConnected;
+    }
+
+    if (request.arguments?['uri'] == null) {
+      return CallToolResult(
+        isError: true,
+        content: [
+          TextContent(text: 'Required parameter "uri" was not provided.'),
+        ],
+      );
+    }
+
+    try {
+      _dtd = await DartToolingDaemon.connect(
+        Uri.parse(request.arguments!['uri'] as String),
+      );
+      unawaited(_dtd!.done.then((_) => _resetDtd()));
+
+      _listenForServices();
+      return CallToolResult(
+        content: [TextContent(text: 'Connection succeeded')],
+      );
+    } catch (e) {
+      return CallToolResult(
+        isError: true,
+        content: [TextContent(text: 'Connection failed: $e')],
+      );
+    }
+  }
+
+  /// Listens to the `Service` stream so we know when the
+  /// `Editor.getDebugSessions` extension method is registered.
+  ///
+  /// The dart tooling daemon must be connected prior to calling this function.
+  void _listenForServices() {
+    final dtd = _dtd!;
+    dtd.onEvent('Service').listen((e) async {
+      switch (e.kind) {
+        case 'ServiceRegistered':
+          if (e.data['service'] == 'Editor' &&
+              e.data['method'] == 'getDebugSessions') {
+            _getDebugSessionsReady = true;
+          }
+        case 'ServiceUnregistered':
+          if (e.data['service'] == 'Editor' &&
+              e.data['method'] == 'getDebugSessions') {
+            _getDebugSessionsReady = false;
+          }
+      }
+    });
+    dtd.streamListen('Service');
+  }
+
+  /// Takes a screenshot of the currently running app.
+  ///
+  /// If more than one debug session is active, then it just uses the first one.
+  //
+  // TODO: support passing a debug session id when there is more than one debug
+  // session.
+  Future<CallToolResult> takeScreenshot(CallToolRequest request) async {
+    final dtd = _dtd;
+    if (dtd == null) return _dtdNotConnected;
+    if (!_getDebugSessionsReady) {
+      // Give it a chance to get ready.
+      await Future<void>.delayed(const Duration(seconds: 1));
+      if (!_getDebugSessionsReady) return _dtdNotReady;
+    }
+
+    final response = await dtd.getDebugSessions();
+    final debugSessions = response.debugSessions;
+    if (debugSessions.isEmpty) return _noActiveDebugSession;
+
+    // TODO: Consider holding on to this connection.
+    final vmService = await vmServiceConnectUri(
+      debugSessions.first.vmServiceUri,
+    );
+
+    try {
+      final vm = await vmService.getVM();
```

Comment:
out of scope for this PR, but we may want to break up which tools are available for a Flutter app vs a pure Dart app.

Reply:
ack